samza-core/src/main/java/org/apache/samza/runtime/ContainerLaunchUtil.java

```diff
@@ -142,19 +150,59 @@ public void beforeStart() {
             public void afterStart() {
               log.info("Container Started");
               listener.afterStart();
+              addShutdownHook();
             }
 
             @Override
             public void afterStop() {
               log.info("Container Stopped");
               listener.afterStop();
+              removeShutdownHook();
+              shutdownLatch.countDown();
             }
 
             @Override
             public void afterFailure(Throwable t) {
               log.info("Container Failed");
               containerRunnerException = t;
               listener.afterFailure(t);
+              removeShutdownHook();
+              shutdownLatch.countDown();
+            }
+
+            private void removeShutdownHook() {
+              try {
+                if (shutdownHookThread != null) {
+                  Runtime.getRuntime().removeShutdownHook(shutdownHookThread);
+                  log.info("Removed Samza container shutdown hook");
+                }
+              } catch (IllegalStateException e) {
+                // Thrown when then JVM is already shutting down, so safe to ignore.
+              }
+            }
+
+            private void addShutdownHook() {
+              shutdownHookThread = new Thread("Samza Container Shutdown Hook Thread") {
+                @Override
+                public void run() {
+                  long shutdownMs = taskConfig.getShutdownMs();
+                  log.info("Attempting to shutdown container from inside shutdownHook, will wait upto {} ms.", shutdownMs);
+                  try {
+                    container.shutdown();
+                    boolean hasShutdown = shutdownLatch.await(shutdownMs, TimeUnit.MILLISECONDS);
+                    if (hasShutdown) {
+                      log.info("Shutdown complete");
+                    } else {
+                      log.error("Did not shut down within {} ms, exiting.", shutdownMs);
+                      ThreadUtil.logThreadDump("Thread dump from Samza Container Shutdown Hook.");
+                    }
+                  } catch (InterruptedException e) {
+                    e.printStackTrace();
```

Comment:
Please use the `log` here instead of `printStackTrace`. Maybe do the same thing here as if it didn't shut down in time.